### ISSUE TITLE
Clearing formatting results in different background color for different renderers

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#4120]
VS Code issue: https://github.com/microsoft/vscode/issues/160497

![image](https://user-images.githubusercontent.com/2193314/190413547-2a7fa930-7784-4f8c-870d-36a059d35e58.png)

The rendering of the repro in the linked issue is different in webgl+canvas vs dom

**Comment by Tyriar** [comment]
Ansi colors test button show a difference too:

webgl:

![image](https://user-images.githubusercontent.com/2193314/209443561-1ff7ea38-e3ff-4efb-a170-befe995c329a.png)


canvas:

![image](https://user-images.githubusercontent.com/2193314/209443567-8ff05c67-4506-44bb-a230-69c49b6589fa.png)

**Comment by benz0li** [comment]
@Tyriar Is there an approximate timeframe for resolving this issue? Is it weeks or rather months?

I want to keep the default setting in code-server; and documenting [workarounds](https://github.com/coder/code-server/pull/6070#issuecomment-1469723499) is not ideal.

**Comment by jerch** [comment]
This is mostly BCE related (also see https://invisible-island.net/ncurses/ncurses.faq.html#bce_mismatches). From my quick tests it seems the canvas and the webgl renderer work properly, but the DOM renderer does not correctly apply BCE to new lines.

Regarding the issue itself:
If BCE is set, the BG color to be used for erasing is not the terminals default BG color anymore, but the currently active one. Now the tricky part - which color should be applied to initialize new lines at the bottom? We currently do that with the active BG color as well, therefore this weird output, if the terminal had to spawn new lines at the end. While thats kinda surprising and differs from output within the initial viewport, it is in fact the expected behavior. From my quick tests only the DOM renderer does not do that correctly, prolly due to early finishing the lines spans (have not investigated).

Did not extensively test this, thus there is a chance I missed some other real bug here. I also cannot repro the differences from the color test above, no clue why.

**Comment by jerch** [comment]
On how to workaround this BCE madness - lets take a simple example:

```bash
echo -e '\x1b[41ma\nb'
```
Intention here is pretty clear - print 2 lines with `a` and `b` with red BG.

A BCE aware TE would print this:
- within original viewport
  ![image](https://user-images.githubusercontent.com/6193135/225305412-20f23fe0-6e1d-497b-aa0c-1c19e8a0d919.png)
- at the bottom creating a new terminal line:
  ![image](https://user-images.githubusercontent.com/6193135/225305876-ec53430e-7f6e-475b-869d-80514d3f1cac.png)
  Due to the need of a new line the currently active BG (red) got applied all over the second line. This raises a few questions:
  - _Why not for the first line?_ Well the first line get created from shell prompt while the default BG color is still active, thus it gets filled with default BG color. Not so for the second line, BG is red at the time of line creation, so the line turns red.
  - _What about the third line?_ The third line (where the new shell prompt is) also does not show the BCE effect for another reason - the PS1 snippet here in use erases the line with defaults before wiring anything. If used with an empty PS1 var, the effect can be triggered (using `sh` to avoid any shell trickery):
  ![image](https://user-images.githubusercontent.com/6193135/225307791-526b35f7-016a-4274-9532-4da04351de85.png)

To avoid the BCE effect, we can explicitly erase the line with BG defaults, either:
- before writing to it (right after the NL)
  ```bash
  echo -e '\x1b[41ma\n\x1b[49m\x1b[2K\x1b[41mb'
  ```
- or by erasing left over cells to the right after writing (before the NL)
  ```bash
  echo -e '\x1b[41ma\nb\x1b[49m\x1b[K'
  ```
Both should give us the wanted output (though keep in mind, that the second variant might destroy output to the right of the cursor, so it is not applicable for complex line manipulations with cursor forth and back moves).

**Comment by benz0li** [comment]
> From my quick tests it seems the canvas and the webgl renderer work properly

Canvas renderer does not work properly.

---

The following settings in code-server (aka VS Code) lead to proper rendering of the terminal:

* `"terminal.integrated.gpuAcceleration": "on"` – if supported
* `"terminal.integrated.gpuAcceleration": "off"`

The following settings lead to improper rendering:

* `"terminal.integrated.gpuAcceleration": "auto"`
* `"terminal.integrated.gpuAcceleration": "canvas"`

---

@jerch: I have whitelisted your GitHub user at https://coder.jupyter.b-data.ch/ so you may have a look at it yourself.

**Comment by benz0li** [comment]
No worries. I am documenting the workaround, then.

**Comment by jerch** [comment]
You mean webgl works properly but canvas does not? Then I am prolly on the wrong track here, for me canvas and webgl behave exactly the same. I wonder if this is a safari only issue (there are some issues with safari at several ends...)

**Comment by benz0li** [comment]
> You mean webgl works properly but canvas does not?

Yes.

> I wonder if this is a safari only issue (there are some issues with safari at several ends...)

No. There is the same issue with the canvas renderer in Firefox.

**Comment by jerch** [comment]
@benz0li Can you give a repro for the issue, how to trigger it?

**Comment by benz0li** [comment]
> @benz0li Can you give a repro for the issue, how to trigger it?

@jerch: I have whitelisted your GitHub user at https://coder.jupyter.b-data.ch/ so you may have a look at it yourself.  
👉 Image: `R (base:test-devtools)`

**Comment by jerch** [comment]
I see the following (in FF):
- webgl
  ![image](https://user-images.githubusercontent.com/6193135/225324508-d5916310-a4c6-4a01-ba7c-27ad4bfab319.png)
- canvas
  ![image](https://user-images.githubusercontent.com/6193135/225324886-77b609df-f16a-4061-9266-2c71c9fb4700.png)
- DOM
  ![image](https://user-images.githubusercontent.com/6193135/225325263-b14b3878-0466-42b2-b4cd-ad8076abc5e5.png)

From he looks only the webgl renderer has the right output atm.

canvas has multiple issues:
- First line of [a, b] should not have turned red.
- There are positioning issues in the powerline prompt.

DOM has one issue not applying BCE at all. Prompt looks right to me (beside minor offset issues, which cannot be solved, it is a font renderer issue of the browser).

So yes there are differences in canvas and webgl. Which xterm.js version is here used? And can you test a different version? (My tests above used 5.1.0, so this either is a regression or got already solved in 5.1.0...)

**Comment by Tyriar** [comment]
> Is there an approximate timeframe for resolving this issue? Is it weeks or rather months?

@benz0li I'm strapped for time at the moment so probably won't be able to look at it for a while, so more like months unless someone else steps in.

**Comment by benz0li** [comment]
> Which xterm.js version is here used?

This issue popped up when [VS Code 1.75.0](https://code.visualstudio.com/updates/v1_75) came out.

xterm.js 5.2.0-beta.21; according to https://github.com/microsoft/vscode/blob/e2816fe719a4026ffa1ee0189dc89bdfdbafb164/package.json#L90

**Comment by jerch** [comment]
Yupp, thats a regression in master branch, can repro it there as well. Hmm.

**Comment by jerch** [comment]
Found a possible culprit: https://github.com/xtermjs/xterm.js/blob/a359c7ca104c66ac1c62ff64b67899b549ef2593/addons/xterm-addon-canvas/src/TextRenderLayer.ts#L99-L102

Commented out these lines, the proper behavior is restored. I introduced these lines in https://github.com/xtermjs/xterm.js/commit/0b56b56c069ab257a9ce55a642d5369dbc04f3ff to greatly speedup the canvas renderer, but was not aware, that it also gets used for BG coloring. This needs a more involved patch to have both good perf and correct behavior, so for now I gonna revert that change --> PR: #4438.
